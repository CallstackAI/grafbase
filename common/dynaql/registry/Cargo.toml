[package]
name = "dynaql-registry"
version = "3.0.31"
authors = ["Grafbase"]
edition = "2021"
description = "Registry Entity for Dynaql"
license = "MIT/Apache-2.0"

[dependencies]
dynaql-value = { path = "../value", version = "3" }
pest = "2"
pest_derive = "2"
serde = { workspace = true }
