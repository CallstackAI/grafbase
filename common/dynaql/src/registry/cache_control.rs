/// Cache control values
///
/// # Examples
///
/// ```rust, ignore
/// use dynaql::*;
///
/// struct Query;
///
/// #[Object(cache_control(max_age = 60))]
/// impl Query {
///     #[graphql(cache_control(max_age = 30))]
///     async fn value1(&self) -> i32 {
///         0
///     }
///
///     #[graphql(cache_control(private))]
///     async fn value2(&self) -> i32 {
///         0
///     }
/// }
///
/// # tokio::runtime::Runtime::new().unwrap().block_on(async {
/// let schema = Schema::new(Query, EmptyMutation, EmptySubscription);
/// assert_eq!(schema.execute("{ value1 }").await.into_result().unwrap().cache_control, CacheControl { public: true, max_age: 30 });
/// assert_eq!(schema.execute("{ value2 }").await.into_result().unwrap().cache_control, CacheControl { public: false, max_age: 60 });
/// assert_eq!(schema.execute("{ value1 value2 }").await.into_result().unwrap().cache_control, CacheControl { public: false, max_age: 30 });
/// # });
/// ```
#[derive(Clone, Copy, PartialEq, Eq, Debug, serde::Deserialize, serde::Serialize, Hash)]
pub struct CacheControl {
    /// Scope is public, default is true.
    pub public: bool,

    /// Cache max age, default is 0.
    pub max_age: usize,

    /// Cache stale_while_revalidate, default is 0.
    #[serde(default)]
    pub stale_while_revalidate: usize,
}

impl Default for CacheControl {
    fn default() -> Self {
        Self {
            public: true,
            max_age: 0,
            stale_while_revalidate: 0,
        }
    }
}

impl CacheControl {
    #[must_use]
    pub(crate) fn merge(self, other: &CacheControl) -> CacheControl {
        CacheControl {
            public: self.public && other.public,
            max_age: if self.max_age == 0 {
                other.max_age
            } else if other.max_age == 0 {
                self.max_age
            } else {
                self.max_age.min(other.max_age)
            },
            stale_while_revalidate: if self.stale_while_revalidate == 0 {
                other.stale_while_revalidate
            } else if other.stale_while_revalidate == 0 {
                self.stale_while_revalidate
            } else {
                self.stale_while_revalidate
                    .min(other.stale_while_revalidate)
            },
        }
    }
}
