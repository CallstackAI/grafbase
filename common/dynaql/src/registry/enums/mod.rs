mod order_by_direction;
pub use order_by_direction::OrderByDirection;

use unindent::unindent;

pub struct DynaqlEnums;

impl DynaqlEnums {
    pub fn sdl() -> String {
        OrderByDirection::sdl()
    }
}

pub trait DynaqlEnum {
    fn ty() -> &'static str;
    fn values() -> Vec<String>;
    fn sdl() -> String {
        unindent(&format!(
            r#"
            enum {ty} {{
                {values}
            }}
            "#,
            ty = Self::ty(),
            values = Self::values().join("\n    ")
        ))
    }
}
