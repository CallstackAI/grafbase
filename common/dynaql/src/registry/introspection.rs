//! Allow converting a [`cynic_introspection::Schema`] into a new [`Registry`].
//!
//! The conversion takes all the known information from the introspection and adds it to the
//! registry.
//!
//! # Panics
//!
//! The conversion is infallible (using [`From`] instead of [`TryFrom`]), but is allowed to panic,
//! if a future update to [`cynic_introspection`] introduces new enum variants to any of the
//! non-exhaustive enums in the crate. This _should_ not happen, unless the actual [GraphQL
//! specification] changes.
//!
//! [GraphQL specification]: https://spec.graphql.org

use super::{
    Deprecation, MetaDirective, MetaEnumValue, MetaField, MetaInputValue, MetaType, Registry,
    __DirectiveLocation,
};

impl From<cynic_introspection::Schema> for Registry {
    fn from(schema: cynic_introspection::Schema) -> Self {
        let mut registry = Self::new();

        // root types
        registry.query_type = schema.query_type;
        registry.mutation_type = schema.mutation_type;
        registry.subscription_type = schema.subscription_type;

        // directives
        registry.directives = schema
            .directives
            .into_iter()
            .map(|d| (d.name.clone(), d.into()))
            .collect();

        // types
        registry.types = schema
            .types
            .into_iter()
            .map(|ty| {
                use cynic_introspection::Type::*;

                let key = match &ty {
                    Object(v) => v.name.clone(),
                    InputObject(v) => v.name.clone(),
                    Enum(v) => v.name.clone(),
                    Interface(v) => v.name.clone(),
                    Union(v) => v.name.clone(),
                    Scalar(v) => v.name.clone(),
                    _ => unimplemented!("unknown graphql type"),
                };

                (key, ty.into())
            })
            .collect();

        registry
    }
}

impl From<cynic_introspection::Directive> for MetaDirective {
    fn from(directive: cynic_introspection::Directive) -> Self {
        Self {
            name: directive.name,
            description: directive.description,
            locations: directive.locations.into_iter().map(Into::into).collect(),
            args: directive
                .args
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            is_repeatable: false,
            visible: None,
        }
    }
}

impl From<cynic_introspection::Field> for MetaField {
    fn from(field: cynic_introspection::Field) -> Self {
        Self {
            name: field.name,
            description: field.description,
            args: field
                .args
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            ty: field.ty.to_string(),
            deprecation: field.deprecated.into(),
            ..Default::default()
        }
    }
}

impl From<cynic_introspection::InputValue> for MetaInputValue {
    fn from(input: cynic_introspection::InputValue) -> Self {
        Self {
            name: input.name,
            description: input.description,
            ty: input.ty.name,
            default_value: input.default_value.map(Into::into),
            visible: None,
            validators: None,
            is_secret: false,
            rename: None,
        }
    }
}

impl From<cynic_introspection::DirectiveLocation> for __DirectiveLocation {
    fn from(location: cynic_introspection::DirectiveLocation) -> Self {
        use __DirectiveLocation::*;
        use cynic_introspection::DirectiveLocation::*;

        match location {
            Query => QUERY,
            Mutation => MUTATION,
            Subscription => SUBSCRIPTION,
            Field => FIELD,
            FragmentDefinition => FRAGMENT_DEFINITION,
            FragmentSpread => FRAGMENT_SPREAD,
            InlineFragment => INLINE_FRAGMENT,
            VariableDefinition => VARIABLE_DEFINITION,
            Schema => SCHEMA,
            Scalar => SCALAR,
            Object => OBJECT,
            FieldDefinition => FIELD_DEFINITION,
            ArgumentDefinition => ARGUMENT_DEFINITION,
            Interface => INTERFACE,
            Union => UNION,
            Enum => ENUM,
            EnumValue => ENUM_VALUE,
            InputObject => INPUT_OBJECT,
            InputFieldDefinition => INPUT_FIELD_DEFINITION,
        }
    }
}

impl From<cynic_introspection::Type> for MetaType {
    fn from(ty: cynic_introspection::Type) -> Self {
        use cynic_introspection::Type::*;

        match ty {
            Object(v) => v.into(),
            InputObject(v) => v.into(),
            Enum(v) => v.into(),
            Interface(v) => v.into(),
            Union(v) => v.into(),
            Scalar(v) => v.into(),
            _ => unimplemented!("unknown graphql type"),
        }
    }
}

impl From<cynic_introspection::ObjectType> for MetaType {
    fn from(object: cynic_introspection::ObjectType) -> Self {
        Self::Object {
            name: object.name,
            description: object.description,
            fields: object
                .fields
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            cache_control: Default::default(),
            extends: false,
            keys: None,
            visible: None,
            is_subscription: false,
            is_node: false,
            rust_typename: String::new(),
            constraints: vec![],
        }
    }
}

impl From<cynic_introspection::Deprecated> for Deprecation {
    fn from(deprecated: cynic_introspection::Deprecated) -> Self {
        use cynic_introspection::Deprecated::*;
        use Deprecation::*;

        match deprecated {
            No => NoDeprecated,
            Yes(reason) => Deprecated { reason },
            _ => unimplemented!("unknown graphql deprecation type"),
        }
    }
}

impl From<cynic_introspection::InputObjectType> for MetaType {
    fn from(input: cynic_introspection::InputObjectType) -> Self {
        Self::InputObject {
            name: input.name,
            description: input.description,
            input_fields: input
                .fields
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            visible: None,
            rust_typename: String::new(),
            oneof: false,
        }
    }
}

impl From<cynic_introspection::EnumType> for MetaType {
    fn from(enum_type: cynic_introspection::EnumType) -> Self {
        Self::Enum {
            name: enum_type.name,
            description: enum_type.description,
            enum_values: enum_type
                .values
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            visible: None,
            rust_typename: String::new(),
        }
    }
}

impl From<cynic_introspection::EnumValue> for MetaEnumValue {
    fn from(enum_value: cynic_introspection::EnumValue) -> Self {
        Self {
            name: enum_value.name,
            description: enum_value.description,
            deprecation: enum_value.deprecated.into(),
            visible: None,
            value: None,
        }
    }
}

impl From<cynic_introspection::InterfaceType> for MetaType {
    fn from(interface: cynic_introspection::InterfaceType) -> Self {
        Self::Interface {
            name: interface.name,
            description: interface.description,
            fields: interface
                .fields
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            possible_types: interface.possible_types.into_iter().collect(),
            extends: false,
            keys: None,
            visible: None,
            rust_typename: String::new(),
        }
    }
}

impl From<cynic_introspection::UnionType> for MetaType {
    fn from(union: cynic_introspection::UnionType) -> Self {
        Self::Union {
            name: union.name,
            description: union.description,
            possible_types: union.possible_types.into_iter().collect(),
            visible: None,
            rust_typename: String::new(),
            discriminators: None,
        }
    }
}

impl From<cynic_introspection::ScalarType> for MetaType {
    fn from(scalar: cynic_introspection::ScalarType) -> Self {
        Self::Scalar {
            name: scalar.name,
            description: scalar.description,
            is_valid: None,
            visible: None,
            specified_by_url: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use cynic_introspection::query::IntrospectionQuery;

    use super::*;

    #[test]
    fn conversion() {
        let data = include_str!("../../tests/swapi_introspection.json");
        let schema = serde_json::from_str::<IntrospectionQuery>(data)
            .unwrap()
            .into_schema()
            .unwrap();

        insta::assert_debug_snapshot!(Registry::from(schema))
    }
}
