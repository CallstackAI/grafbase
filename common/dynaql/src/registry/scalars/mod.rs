use std::str::FromStr;

use chrono::{DateTime, Utc};
use const_format::concatcp;
use dynaql_value::ConstValue;
use fast_chemail::parse_email;
use url::Url;

use crate::{InputValueError, InputValueResult};

/// The `PossibleScalar` enum is the list of possible Scalar usable within dynaql
#[derive(Debug, Clone, Copy)]
pub enum PossibleScalar {
    String,
    Int,
    Float,
    Boolean,
    ID,
    DateTime,
    JSON,
    Url,
    Email,
}

const SPECIFIED_BY_DIRECTIVE: &str = r#"
directive @specifiedBy(url: String!) on SCALAR
"#;

const DATETIME_DIRECTIVE: &str = r#"
"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, is compliant with the date-time format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.

This scalar is a description of an exact instant on the timeline such as the instant that a user account was created.
"""
scalar DateTime @specifiedBy(url: "https://datatracker.ietf.org/doc/html/rfc3339")
"#;

const URL_DIRECTIVE: &str = r#"
"""
A URL as defined byÂ RFC1738. For example, `https://grafbase.com/foo/` or `mailto:example@grafbase.com`.
"""
scalar URL @specifiedBy(url: "http://url.spec.whatwg.org/")
"#;

const JSON_DIRECTIVE: &str = r#"
"""
Any JSON value.
"""
scalar JSON
"#;

const EMAIL_DIRECTIVE: &str = r#"
"""
A scalar to validate the email as it is defined in the HTML specification.
"""
scalar Email @specifiedBy(url: "https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address")
"#;

impl PossibleScalar {
    /// Function to **check** if the inputed value is able to be cast into the expected type.
    /// TODO: In the future, we should also do more than just a check at the request parsing, we
    /// should also allow casting to an expected scalar to be used inside the resolving chain.
    pub(crate) fn check_valid(&self, value: &ConstValue) -> InputValueResult<bool> {
        match (self, value) {
            (Self::String | Self::ID, ConstValue::String(_)) => Ok(true),
            (Self::Boolean, ConstValue::Boolean(_)) => Ok(true),
            (Self::Int, ConstValue::Number(num)) => Ok(!num.is_f64()),
            (Self::Float, ConstValue::Number(_)) => Ok(true),
            (Self::DateTime, ConstValue::String(date)) => {
                date.parse::<DateTime<Utc>>()
                    .map_err(|err| InputValueError::ty_custom("DateTime", err))?;
                Ok(true)
            }
            (Self::Url, ConstValue::String(value)) => Url::from_str(value)
                .map_err(|err| InputValueError::ty_custom("URL", err))
                .map(|_| true),
            (Self::Email, ConstValue::String(value)) => parse_email(value)
                .map_err(|err| InputValueError::ty_custom("Email", err))
                .map(|_| true),
            (Self::JSON, ConstValue::Object(_)) => Ok(true),
            _ => Ok(false),
        }
    }

    /// Generate directives associated
    pub const fn directives() -> &'static str {
        concatcp!(
            SPECIFIED_BY_DIRECTIVE,
            '\n',
            DATETIME_DIRECTIVE,
            '\n',
            JSON_DIRECTIVE,
            '\n',
            URL_DIRECTIVE,
            '\n',
            EMAIL_DIRECTIVE
        )
    }
}

#[derive(Debug, Clone, thiserror::Error)]
pub enum PossibleScalarErrors {
    #[error("\"{expected_ty}\" is not a proper scalar")]
    NotAScalar { expected_ty: String },
}

impl TryFrom<&str> for PossibleScalar {
    type Error = PossibleScalarErrors;
    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "String" => Ok(PossibleScalar::String),
            "Int" => Ok(PossibleScalar::Int),
            "Float" => Ok(PossibleScalar::Float),
            "Boolean" => Ok(PossibleScalar::Boolean),
            "ID" => Ok(PossibleScalar::ID),
            "DateTime" => Ok(PossibleScalar::DateTime),
            "JSON" => Ok(PossibleScalar::JSON),
            "URL" => Ok(PossibleScalar::Url),
            "Email" => Ok(PossibleScalar::Email),
            _ => Err(PossibleScalarErrors::NotAScalar {
                expected_ty: value.to_string(),
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::PossibleScalar;
    use dynaql_value::ConstValue;

    #[test]
    fn check_json_valid() {
        let value = serde_json::json!({
            "code": 200,
            "success": true,
            "payload": {
                "features": [
                    "serde",
                    "json"
                ]
            }
        });

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::JSON.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_url_valid() {
        let value = serde_json::Value::String("https://grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Url.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_url_mailto_valid() {
        let value = serde_json::Value::String("mailto:anthony@grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Url.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_mail_valid() {
        let value = serde_json::Value::String("anthony@grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Email.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_not_mail_valid() {
        let value = serde_json::Value::String("anthony @grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Email.check_valid(&const_value);
        assert!(scalar.is_err());
    }
}
