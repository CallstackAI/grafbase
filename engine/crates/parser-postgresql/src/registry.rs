use postgresql_types::database_definition::DatabaseDefinition;
use parser_sdl::Registry;

pub(super) fn generate(database_definition: DatabaseDefinition, name: &str, _namespaced: bool) -> Registry {
    let mut registry = Registry::default();

    registry
        .postgres_databases
        .insert(name.to_string(), database_definition);

    registry
}
