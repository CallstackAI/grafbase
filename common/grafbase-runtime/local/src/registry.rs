use async_trait::async_trait;
use grafbase_runtime::registry::{RegistryError, RegistryProvider, RegistryResult};
use worker::Env;

pub struct Registry<'a> {
    env: &'a Env,
    var_name: &'a str,
}

impl<'a> Registry<'a> {
    pub fn new(env: &'a Env, var_name: &'a str) -> Self {
        Registry { env, var_name }
    }
}

#[async_trait(?Send)]
impl RegistryProvider for Registry<'_> {
    async fn get_registry(&self) -> RegistryResult<worker::wasm_bindgen::JsValue> {
        use worker_utils::{EnvExt, VarType};
        let registry: String = self
            .env
            .var_get(VarType::Var, self.var_name)
            .map_err(|_err| RegistryError::MissingEnvVar(self.var_name.to_string()))?;
        Ok(worker::js_sys::JSON::parse(&registry).unwrap())
    }
}
