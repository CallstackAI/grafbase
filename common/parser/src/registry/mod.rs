//! TODO:
//!
//! -> Split each of the creation and add tests with SDL
//!
use case::CaseExt;

use dynaql::indexmap::IndexMap;
use dynaql::registry::enums::DynaqlEnum;
use dynaql::registry::relations::MetaRelation;
use dynaql::registry::transformers::Transformer;
use dynaql::registry::{
    resolvers::dynamo_mutation::DynamoMutationResolver, resolvers::Resolver, resolvers::ResolverType,
    variables::VariableResolveDefinition, MetaField, MetaInputValue, MetaType,
};
use dynaql::registry::{MetaEnumValue, Registry};
use dynaql::validation::dynamic_validators::DynValidator;
use dynaql::AuthConfig;
use dynaql_parser::types::{BaseType, FieldDefinition, ObjectType, TypeDefinition};
use grafbase::auth::Operations;
use std::fmt::Display;

use crate::registry::names::MetaNames;
use crate::rules::length_directive::{LENGTH_DIRECTIVE, MAX_ARGUMENT, MIN_ARGUMENT};
use crate::rules::visitor::VisitorContext;
use crate::utils::{to_input_type, to_lower_camelcase};

mod mutations;
pub mod names;
mod pagination;
mod relations;
mod search;
pub use mutations::{add_mutation_create, add_mutation_update, NumericFieldKind};
pub use pagination::{add_query_paginated_collection, generate_pagination_args};
pub use search::add_query_search;

fn register_dynaql_enum<T: DynaqlEnum>(registry: &mut Registry) -> BaseType {
    let type_name = T::ty().to_string();
    registry.create_type(
        |_| MetaType::Enum {
            name: type_name.clone(),
            description: None,
            enum_values: IndexMap::from_iter(
                T::values()
                    .into_iter()
                    .map(|value| {
                        (
                            value.clone(),
                            MetaEnumValue {
                                name: value,
                                description: None,
                                deprecation: dynaql::registry::Deprecation::NoDeprecated,
                                visible: None,
                                value: None,
                            },
                        )
                    })
                    .collect::<Vec<_>>(),
            ),
            visible: None,
            rust_typename: type_name.clone(),
        },
        &type_name,
        &type_name,
    );
    BaseType::named(&type_name)
}

/// Create an input type for a non_primitive Type.
pub fn add_input_type_non_primitive(ctx: &mut VisitorContext<'_>, object: &ObjectType, type_name: &str) -> String {
    let type_name = type_name.to_string();
    let input_type = format!("{}Input", type_name.to_camel());

    // Input
    ctx.registry.get_mut().create_type(
        |_| MetaType::InputObject {
            name: input_type.clone(),
            description: Some(format!("{type_name} input type.")),
            oneof: false,
            input_fields: {
                let mut input_fields = IndexMap::new();
                for field in &object.fields {
                    let name = &field.node.name.node;

                    input_fields.insert(
                        name.clone().to_string(),
                        MetaInputValue {
                            description: field.node.description.clone().map(|x| x.node),
                            ..MetaInputValue::new(
                                name.to_string(),
                                to_input_type(&ctx.types, field.node.ty.clone().node),
                            )
                        },
                    );
                }
                input_fields
            },
            visible: None,
            rust_typename: input_type.clone(),
        },
        &input_type,
        &input_type,
    );

    input_type
}

/// Add the remove mutation for a given Object
pub fn add_remove_mutation(ctx: &mut VisitorContext<'_>, type_name: &str, auth: Option<&AuthConfig>) {
    let type_name = type_name.to_string();
    let delete_payload_name = format!("{}DeletePayload", type_name.to_camel());

    // DeletePayload
    ctx.registry.get_mut().create_type(
        |_| MetaType::Object {
            name: delete_payload_name.clone(),
            description: None,
            fields: {
                let mut fields = IndexMap::new();
                let name = "deletedId".to_string();
                fields.insert(
                    name.clone(),
                    MetaField {
                        name,
                        description: None,
                        args: Default::default(),
                        // TODO: Should be infered from the entity depending on the directives
                        ty: "ID!".to_string(),
                        deprecation: Default::default(),
                        cache_control: Default::default(),
                        external: false,
                        requires: None,
                        provides: None,
                        visible: None,
                        compute_complexity: None,
                        edges: Vec::new(),
                        relation: None,
                        resolve: None,
                        transformer: Some(Transformer::JSONSelect {
                            property: "id".to_string(),
                        }),
                        plan: None,
                        required_operation: Some(Operations::DELETE),
                        auth: auth.cloned(),
                    },
                );
                fields
            },
            cache_control: Default::default(),
            extends: false,
            keys: None,
            is_node: false,
            visible: None,
            is_subscription: false,
            rust_typename: delete_payload_name.clone(),
            constraints: vec![],
        },
        &delete_payload_name,
        &delete_payload_name,
    );

    // deleteMutation
    ctx.mutations.push(MetaField {
        name: format!("{}Delete", to_lower_camelcase(&type_name)),
        description: Some(format!("Delete a {type_name} by ID or unique field")),
        args: {
            let mut args = IndexMap::new();
            args.insert(
                "by".to_owned(),
                MetaInputValue::new("by", format!("{type_name}ByInput!")),
            );
            args
        },
        ty: delete_payload_name,
        deprecation: dynaql::registry::Deprecation::NoDeprecated,
        cache_control: Default::default(),
        external: false,
        provides: None,
        requires: None,
        visible: None,
        edges: Vec::new(),
        relation: None,
        compute_complexity: None,
        resolve: Some(Resolver {
            id: Some(format!("{}_delete_resolver", type_name.to_lowercase())),
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::DeleteNode {
                ty: type_name,
                by: VariableResolveDefinition::InputTypeName("by".to_owned()),
            }),
        }),
        plan: None,
        transformer: None,
        required_operation: Some(Operations::DELETE),
        auth: auth.cloned(),
    });
}

fn get_length_validator(field: &FieldDefinition) -> Option<DynValidator> {
    use tuple::Map;
    field
        .directives
        .iter()
        .find(|directive| directive.node.name.node == LENGTH_DIRECTIVE)
        .map(|directive| {
            let (min_value, max_value) = (MIN_ARGUMENT, MAX_ARGUMENT).map(|argument_name| {
                directive.node.get_argument(argument_name).and_then(|argument| {
                    if let dynaql_value::ConstValue::Number(ref min) = argument.node {
                        min.as_u64().and_then(|min| min.try_into().ok())
                    } else {
                        None
                    }
                })
            });
            DynValidator::length(min_value, max_value)
        })
}

/// Used to keep track of the parent relation when created nested input types
/// TODO: Merge it with MetaRelation?
pub struct ParentRelation<'a> {
    /// TypeDefinition of @model type
    model_type_definition: &'a TypeDefinition,
    meta: &'a MetaRelation,
}

impl<'a> Display for ParentRelation<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} relation of {}",
            self.meta.name,
            MetaNames::model(self.model_type_definition)
        )
    }
}
