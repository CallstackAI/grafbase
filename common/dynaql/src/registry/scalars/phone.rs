use super::{DynamicParse, SDLDefinitionScalar};
use crate::{Error, InputValueError, InputValueResult};
use dynaql_value::ConstValue;
use phonenumber as _;

// TODO: Input coercion to accept either ms or a date
pub struct PhoneScalar;

impl<'a> SDLDefinitionScalar<'a> for PhoneScalar {
    fn name() -> Option<&'a str> {
        Some("Phone")
    }

    fn description() -> Option<&'a str> {
        Some("A phone number. This value is stored as a string. Phone numbers can contain either spaces or hyphens to separate digit groups. Phone numbers without a country code are assumed to be US/North American numbers adhering to theÂ North American Numbering Plan (NANP).")
    }
}

impl DynamicParse for PhoneScalar {
    fn is_valid(value: &ConstValue) -> bool {
        match value {
            ConstValue::String(phone) => phonenumber::parse(None, phone)
                .map(|phone| phone.is_valid())
                .unwrap_or_default(),
            _ => false,
        }
    }

    fn to_value(value: serde_json::Value) -> Result<ConstValue, Error> {
        match value {
            serde_json::Value::String(phone) => Ok(ConstValue::String(phone)),
            _ => Err(Error::new(
                "Data violation: Cannot coerce the initial value into an Phone",
            )),
        }
    }

    fn parse(value: ConstValue) -> InputValueResult<serde_json::Value> {
        match value {
            ConstValue::String(phone) => {
                let e164_phone = phonenumber::parse(None, phone)
                    .map_err(|err| InputValueError::ty_custom("Phone", err))?;

                if e164_phone.is_valid() {
                    Ok(serde_json::Value::String(
                        e164_phone
                            .format()
                            .mode(phonenumber::Mode::International)
                            .to_string(),
                    ))
                } else {
                    Err(InputValueError::ty_custom("Phone", "Invalid phone number"))
                }
            }
            _ => Err(InputValueError::ty_custom(
                "Phone",
                "Cannot parse into a Phone",
            )),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::super::SDLDefinitionScalar;
    use crate::registry::scalars::{DynamicParse, PhoneScalar};
    use dynaql_value::ConstValue;
    use insta::assert_snapshot;

    #[test]
    fn check_test_phonenumber() {
        let value = serde_json::Value::String("+33612121212".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PhoneScalar::parse(const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_test_fail_phonenumber() {
        let value = serde_json::Value::String("+3361212121".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PhoneScalar::parse(const_value);
        assert!(scalar.is_err());
        insta::assert_debug_snapshot!(scalar);
    }

    #[test]
    fn ensure_directives_sdl() {
        assert_snapshot!(PhoneScalar::sdl());
    }
}
