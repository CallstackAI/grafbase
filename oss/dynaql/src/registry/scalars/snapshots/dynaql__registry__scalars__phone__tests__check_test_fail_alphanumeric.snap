---
source: common/dynaql/src/registry/scalars/phone.rs
assertion_line: 115
expression: scalar
---
Err(
    InputValueError {
        message: "Failed to parse \"PhoneNumber\": Invalid phone number",
        phantom: PhantomData<serde_json::value::Value>,
    },
)
